nomos-da/network/core/src/protocols/replication/behaviour.rs

```diff
@@ -156,361 +218,254 @@ where
-            Either::Left(e) => e,
-            Either::Right(v) => libp2p::core::util::unreachable(v),
-        };
-        match event {
-            HandlerEventToBehaviour::IncomingMessage { message } => {
-                self.replicate_message(message.clone());
-                self.outgoing_events.push_back(ToSwarm::GenerateEvent(
-                    ReplicationEvent::IncomingMessage {
-                        peer_id,
-                        message: Box::new(message),
-                    },
-                ));
-            }
-            HandlerEventToBehaviour::OutgoingMessageError { error } => {
-                self.outgoing_events.push_back(ToSwarm::GenerateEvent(
-                    ReplicationError::Io { peer_id, error }.into(),
-                ));
-            }
-        }
-        self.try_wake();
+        let Either::Left(event) = event;
+        self.stream_behaviour
+            .on_connection_handler_event(peer_id, connection_id, event)
     }
 
     fn poll(
         &mut self,
         cx: &mut Context<'_>,
     ) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
-        if let Some(event) = self.outgoing_events.pop_front() {
-            Poll::Ready(event)
-        } else {
-            self.waker = Some(cx.waker().clone());
-            Poll::Pending
-        }
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use std::{
-        collections::HashSet,
-        sync::Arc,
-        task::{Context, Poll},
-    };
-
-    use futures::task::{waker_ref, ArcWake};
-    use kzgrs_backend::testutils::get_da_blob;
-    use libp2p::{identity, PeerId};
-    use nomos_core::da::BlobId;
-    use nomos_da_messages::common::Blob;
-
-    use super::*;
-
-    #[derive(Clone, Debug)]
-    struct MockMembershipHandler {
-        membership: HashMap<PeerId, HashSet<SubnetworkId>>,
-    }
-
-    impl MembershipHandler for MockMembershipHandler {
-        type NetworkId = SubnetworkId;
-        type Id = PeerId;
-
-        fn membership(&self, peer_id: &PeerId) -> HashSet<Self::NetworkId> {
-            self.membership.get(peer_id).cloned().unwrap_or_default()
-        }
+        let mut should_wake = false;
 
-        fn is_allowed(&self, _id: &Self::Id) -> bool {
-            unimplemented!()
+        // Check if we've received any new messages
+        match self.incoming_tasks.poll_next_unpin(cx) {
+            Poll::Ready(Some(Ok((peer_id, message, stream)))) => {
+                // Replicate the message to all connected peers from the same subnet if we
+                // haven't seen it yet
+                self.replicate_message(message.clone());
+                // Wait for any next incoming message on the same stream
+                self.incoming_tasks
+                    .push(Self::try_read_message(peer_id, stream).boxed());
+                // Signal to the swarm that we've received the message
+                return Poll::Ready(ToSwarm::GenerateEvent(ReplicationEvent::IncomingMessage {
```

Comment:
I know this is WIP, but if we return here, behaviour will only start processing outgoing tasks after all incoming tasks are finished - in theory we might never process outgoing tasks. I would recommend to only process one outgoing + one incoming task during the poll.